File: src/editor/serialization/legacy/FilterSerializer.java
package editor.serialization.legacy;

import java.io.IOException;
import java.io.ObjectInputStream;

import editor.database.card.CardLayout;
import editor.database.characteristics.Expansion;
import editor.database.characteristics.ManaCost;
import editor.database.characteristics.ManaType;
import editor.database.characteristics.Rarity;
import editor.filter.Filter;
import editor.filter.FilterAttribute;
import editor.filter.FilterGroup;
import editor.filter.leaf.ColorFilter;
import editor.filter.leaf.ManaCostFilter;
import editor.filter.leaf.NumberFilter;
import editor.filter.leaf.TextFilter;
import editor.filter.leaf.TypeLineFilter;
import editor.filter.leaf.options.multi.MultiOptionsFilter;
import editor.filter.leaf.options.single.ExpansionFilter;
import editor.filter.leaf.options.single.LayoutFilter;
import editor.filter.leaf.options.single.RarityFilter;
import editor.util.Comparison;
import editor.util.Containment;

public interface FilterSerializer
{
    public static Filter readExternal(ObjectInputStream in) throws IOException, ClassNotFoundException
    {
        FilterAttribute type = FilterAttribute.fromString(in.readUTF());
        if (type == FilterAttribute.GROUP)
        {
            FilterGroup filter = new FilterGroup();
            filter.mode = (FilterGroup.Mode)in.readObject();
            int n = in.readInt();
            for (int i = 0; i < n; i++)
                filter.addChild(readExternal(in));
            return filter;
        }
        else
        {
            int n = 0;
            switch (type)
            {
            case NAME:
            case RULES_TEXT:
            case FLAVOR_TEXT:
            case PRINTED_TEXT:
            case ARTIST:
            case PRINTED_TYPES:
                TextFilter text = (TextFilter)type.get();
                text.contain = (Containment)in.readObject();
                text.regex = in.readBoolean();
                text.text = in.readUTF();
                return text;
            case LAYOUT:
                LayoutFilter layout = (LayoutFilter)type.get();
                layout.contain = (Containment)in.readObject();
                n = in.readInt();
                for (int i = 0; i < n; i++)
                {
                    in.readBoolean(); // Should be true, since CardLayout is serializable
                    layout.selected.add((CardLayout)in.readObject());
                }
                return layout;
            case MANA_COST:
                ManaCostFilter mana = (ManaCostFilter)type.get();
                mana.contain = (Containment)in.readObject();
                mana.cost = ManaCost.parseManaCost(in.readUTF());
                return mana;
            case CMC:
            case CARD_NUMBER:
                NumberFilter number = (NumberFilter)type.get();
                number.operand = in.readDouble();
                number.operation = (Comparison)in.readObject();
                return number;
            case COLOR:
            case COLOR_IDENTITY:
                ColorFilter color = (ColorFilter)type.get();
                color.contain = (Containment)in.readObject();
                n = in.readInt();
                for (int i = 0; i < n; i++)
                    color.colors.add((ManaType)in.readObject());
                color.multicolored = in.readBoolean();
                return color;
            case TYPE_LINE:
                TypeLineFilter line = (TypeLineFilter)type.get();
                line.contain = (Containment)in.readObject();
                line.line = in.readUTF();
                return line;
            case SUPERTYPE:
            case CARD_TYPE:
            case SUBTYPE:
            case BLOCK:
                @SuppressWarnings("unchecked")
                MultiOptionsFilter<String> string = (MultiOptionsFilter<String>)type.get();
                string.contain = (Containment)in.readObject();
                n = in.readInt();
                for (int i = 0; i < n; i++)
                {
                    in.readBoolean(); // Should be true since strings are serializable
                    string.selected.add((String)in.readObject());
                }
                return string;
            case EXPANSION:
                ExpansionFilter expansion = (ExpansionFilter)type.get();
                expansion.contain = (Containment)in.readObject();
                n = in.readInt();
                for (int i = 0; i < n; i++)
                {
                    in.readBoolean(); // Should be false since Expansions are not serializable
                    String name = in.readUTF();
                    for (Expansion e : Expansion.expansions)
                    {
                        if (name.equalsIgnoreCase(e.name))
                        {
                            expansion.selected.add(e);
                            break;
                        }
                    }
                }
                return expansion;
            case RARITY:
                RarityFilter rarity = (RarityFilter)type.get();
                rarity.contain = (Containment)in.readObject();
                n = in.readInt();
                for (int i = 0; i < n; i++)
                {
                    in.readBoolean(); // Should be false, since Raritys are not serializable
                    rarity.selected.add(Rarity.parseRarity(in.readUTF()));
                }
                return rarity;
            case POWER:
            case TOUGHNESS:
            case LOYALTY:
                return null;
            case FORMAT_LEGALITY:
                return null;
            case TAGS:
                return null;
            case NONE:
                return null;
            case ANY:
                return null;
            case DEFAULTS: // Shouldn't actually show up
            default:
                return null;
            }
        }
    }    
}

File: src/main/scala/editor/serialization/FilterSerializer.scala
package editor.serialization

import editor.database.attributes.CardAttribute
import editor.database.attributes.Expansion
import editor.database.attributes.ManaCost
import editor.database.attributes.ManaType
import editor.database.attributes.Rarity
import editor.database.card.CardLayout
import editor.filter.FaceSearchOptions
import editor.filter.Filter
import editor.filter.FilterGroup
import editor.filter.leaf._
import editor.util.Comparison
import editor.util.Containment
import org.json4s._
import org.json4s.native._
import editor.database.attributes.HasTextFilter
import editor.database.attributes.HasNumberFilter
import editor.database.attributes.HasColorFilter
import editor.database.attributes.HasSingletonOptionsFilter

/**
 * JSON serializer/deserializer for [[Filter]]s using their methods for converting to/from JSON objects.
 * @author Alec Roelke
 */
object FilterSerializer extends CustomSerializer[Filter](implicit format => (
  { case v =>
    val faces = (v \ "faces").extract[Option[String]].map(FaceSearchOptions.valueOf).getOrElse(FaceSearchOptions.ANY)
    val selected = (v \ "selected").extract[Option[Set[String]]]
    val contain = (v \ "contains").extract[Option[String]].flatMap(Containment.parse)
    (v \ "type").extract[CardAttribute[?, ?]] match {
      case text: HasTextFilter => text.filter.copy(faces = faces)
      case number: HasNumberFilter => number.filter.copy(faces = faces)
      case color: HasColorFilter => color.filter
      case single: HasSingletonOptionsFilter[_] => single.filter.copy(selected = selected.map(_.flatMap(single.parse)).get)
      case CardAttribute.ManaCost => CardAttribute.ManaCost.filter.copy(faces = faces)
      case CardAttribute.Devotion => CardAttribute.Devotion.filter.copy(faces = faces, types = (v \ "colors") match {
        case JArray(colors) => colors.map((s) => ManaType.parse(s.extract[String]).get).toSet
        case x => throw MatchError(x)
      }, operation = Comparison.valueOf((v \ "operation").extract[String].apply(0)), operand = (v \ "operand").extract[Int])
      case CardAttribute.TypeLine => CardAttribute.TypeLine.filter.copy(faces = faces)
      case CardAttribute.CardType => CardAttribute.CardType.filter.copy(selected = selected.get)
      case CardAttribute.Subtype => CardAttribute.Subtype.filter.copy(selected = selected.get)
      case CardAttribute.Supertype => CardAttribute.Supertype.filter.copy(selected = selected.get)
      case CardAttribute.LegalIn => CardAttribute.LegalIn.filter.copy(selected = selected.get, restricted = (v \ "restricted").extract[Boolean])
      case CardAttribute.Tags => CardAttribute.Tags.filter.copy(selected = selected.get)
      case CardAttribute.AnyCard => CardAttribute.AnyCard.filter
      case CardAttribute.NoCard => CardAttribute.NoCard.filter
      case CardAttribute.Group => FilterGroup(
        (v \ "children").extract[Seq[Filter]],
        FilterGroup.Mode.values.find(_.toString.equalsIgnoreCase((v \ "mode").extract[String])).getOrElse(FilterGroup.Mode.And),
        (v \ "comment").extract[Option[String]].getOrElse("")
      )
      case x => throw IllegalArgumentException(s"$x is not a filterable attribute")
    } match {
      case t: TextFilter => t.copy(contain = contain.get, regex = (v \ "regex").extract[Boolean], text = (v \ "pattern").extract[String])
      case t: TypeLineFilter => t.copy(contain = contain.get, line = (v \ "pattern").extract[String])
      case m: ManaCostFilter => m.copy(contain = contain.get, cost = ManaCost.parse((v \ "cost").extract[String]).get)
      case c: ColorFilter =>
        val colors = (v \ "colors") match {
          case JArray(colors) => colors.map((s) => ManaType.parse(s.extract[String]).get).toSet
          case x => throw MatchError(x)
        }
        c.copy(contain = contain.get, colors = colors, multicolored = (v \ "multicolored").extract[Boolean])
      case n: NumberFilter => n.copy(operation = Comparison.valueOf((v \ "operation").extract[String].apply(0)), operand = (v \ "operand").extract[Double], varies = n.variable.isDefined && (v \ "varies").extract[Boolean])
      case s: SingletonOptionsFilter[?] => s.copy(contain = contain.get)
      case m: MultiOptionsFilter[?] => m.copy(contain = contain.get)
      case f => f
    }},
  { case filter: Filter => JObject(List(JField("type", JString(filter.attribute.toString))) ++ (filter match {
    case l: FilterLeaf => List(JField("faces", JString(l.faces.toString)))
    case g: FilterGroup => List(
      JField("mode", JString(g.mode.toString)),
      JField("comment", JString(g.comment)),
      JField("children", JArray(g.map(Extraction.decompose).toList))
    )
  }) ++ (filter match {
    case o: OptionsFilter[?, ?] => List(JField("contains", JString(o.contain.toString)))
    case _ => Nil
  }) ++ (filter match {
    case t: TypeLineFilter => List(
      JField("contains", JString(t.contain.toString)),
      JField("pattern", JString(t.line))
    )
    case t: TextFilter => List(
      JField("contains", JString(t.contain.toString)),
      JField("regex", JBool(t.regex)),
      JField("pattern", JString(t.text))
    )
    case n: NumberFilter => List(
      JField("operation", JString(n.operation.toString)),
      JField("operand", JDouble(n.operand))
    ) ++ n.variable.map(_ => JField("varies", JBool(n.varies))).toList
    case m: ManaCostFilter => List(
      JField("contains", JString(m.contain.toString)),
      JField("cost", JString(m.cost.toString))
    )
    case c: ColorFilter => List(
      JField("contains", JString(c.contain.toString)),
      JField("colors", JArray(c.colors.map((m) => JString(m.toString)).toList)),
      JField("multicolored", JBool(c.multicolored))
    )
    case d: DevotionFilter => List(
      JField("colors", JArray(d.types.map((m) => JString(m.toString)).toList)),
      JField("operation", JString(d.operation.toString)),
      JField("operand", JInt(d.operand))
    )
    case _: BinaryFilter => Nil // Nothing additional actually needs to be serialized
    case o: OptionsFilter[?, ?] => List(JField("selected", JArray(o.selected.map((e) => JString(e.toString)).toList)))
    case _ => Nil
  }))}
))